[KafkaIO] Fix average record size data race and backlog estimation

sjvanrossum opened the pull request:
The offset gap ratio may artificially shrink the backlog if consumers can't catch up to the tail of an expiring topic. This may cause runners to trigger a downscaling event which worsens the issue.

MovingAvg has been modified to atomically write the accumulated state since concurrent normal loads/stores of longs/doubles may tear. The numUpdates field is only used by the writer and can be kept as non-volatile, but the update method ensures that normal loads/stores on numUpdates are ordered in relation to acquiring loads and releasing stores on avg. To prevent false sharing I've padded the class since there may be tens to hundreds of instances of the accumulator and updates happen per consumed record.

The JMH benchmark I've added shows a slight uplift in average time per op for both reads and writes compared to the current implementation.

Results of task `:sdks:java:io:kafka:jmh:jmh` on a t2d-standard-60 Cloud Workstation:
```
Benchmark                                 Mode  Cnt       Score       Error  Units
KafkaIOUtilsBenchmark.Atomic              avgt   15   50693.751 ±  4937.770  ns/op
KafkaIOUtilsBenchmark.Atomic:atomicRead   avgt   15    5577.357 ±  1135.962  ns/op
KafkaIOUtilsBenchmark.Atomic:atomicWrite  avgt   15  140926.539 ± 14847.095  ns/op
KafkaIOUtilsBenchmark.Plain               avgt   15   65018.754 ±  9814.457  ns/op
KafkaIOUtilsBenchmark.Plain:plainRead     avgt   15    6658.736 ±   288.912  ns/op
KafkaIOUtilsBenchmark.Plain:plainWrite    avgt   15  181738.789 ± 29403.883  ns/op
```

Note that this test likely does not highlight the effect of padding since it doesn't construct a large pool of accumulators.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

sjvanrossum commented:
Run Spotless PreCommit

kennknowles commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIOUtils.java:
```diff
@@ -129,19 +131,80 @@ static Map<String, Object> getOffsetConsumerConfig(
+   */
+  @SuppressFBWarnings("UUF_UNUSED_FIELD")
+  private static class MovingAvgPadding {
+    byte p000, p001, p002, p003, p004, p005, p006, p007;
```
Was this a reaction to a measured slowdown? I'm not like super opposed to having something like this, but I expect it's relevance and effectiveness to wane as the codebase evolves, unless we have some fairly deep and clear understanding of exactly what needs to be maintained.

sjvanrossum replied:
Removed layout padding, see comment below.

kennknowles commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIOUtils.java:
```diff
@@ -129,19 +131,80 @@ static Map<String, Object> getOffsetConsumerConfig(
+
+  // The accumulator's fields should be padded to at least 128 bytes (at least 1 or 2
+  // cache lines).
+  private static class MovingAvgFields extends MovingAvgPadding {
```
Why is the padding not just inlined here? It would be clearer what is going on and saves a layer of inheritance.

sjvanrossum replied:
The compiler may rearrange fields within a class arbitrarily, but it can't freely rearrange fields within a class hierarchy except to fill alignment gaps in a super class since that space would be unused in an instance of the super class and thus it's safe for any subclass to place its fields there (as happens in JDK 15 and up).

Example:
```
// Fields may be rearranged in any order by the compiler.
class A {
  long prePaddingField, ...;
  volatile long importantField;
  long postPaddingField, ...;
}

// For compactness use 8 byte fields instead to allow the compiler to fill alignment gaps
class APrePadding {
  long prePaddingField, ...;
}

class AFields extends APrePadding {
  volatile long importantField;
}

// For compactness use 8 byte fields instead to allow the compiler to fill alignment gaps
class APostPadding extends AFields {
  long postPaddingField, ...;
}

// This class should have prePaddingField(s) placed before importantField and postPaddingField(s) placed after importantField
class PaddedA extends APostPadding {}
```

Padding before the volatile field should ensure it is loaded onto a different cache line than the mark word of the object header and padding after the volatile field should ensure the same for other objects in the cache.
See https://shipilev.net/jvm/objects-inside-out/#_field_layout_across_the_hierarchy and https://github.com/ben-manes/caffeine/blob/v3.2.0/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedBuffer.java#L172 for examples and https://github.com/google/guava/blame/v33.4.0/android/guava/src/com/google/common/hash/Striped64.java#L103 in Guava for a comment about field reordering. Separation of fields through the class hierarchy is considered more robust against reordering, but given that Guava doesn't use that pattern it would seem it doesn't matter much in practice.

In any case, after adjusting the benchmark experiments to model the interaction between readers and writers that I expect in Beam it appears that layout padding is irrelevant. :)

kennknowles commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIOUtils.java:
```diff
@@ -129,19 +131,80 @@ static Map<String, Object> getOffsetConsumerConfig(
+   * Sanity of visibility is only useful when the writer thread changes since avg is the only field that can be shared between multiple concurrent threads.
+   */
+  @SuppressFBWarnings("UUF_UNUSED_FIELD")
+  public static final class MovingAvg extends MovingAvgFields {
```
This class should simply have a private `MovingAvgFields` member.

sjvanrossum replied:
Removed layout padding, see comment below.

kennknowles commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIOUtils.java:
```diff
@@ -129,19 +131,80 @@ static Map<String, Object> getOffsetConsumerConfig(
+      return Double.longBitsToDouble(avg);
+    }
+
+    protected void setAvg(final double value) {
```
Just make these package-private and call them from the MovingAvg class. (see below: use a field instead of inheritance)

sjvanrossum replied:
Removed layout padding, see comment below.

sjvanrossum commented:
@kennknowles I've updated the experiment setup in the benchmark. I've reduced concurrent readers to 1, because as far as I know only the `ProcessElement` thread reads and writes the accumulator and a `GetSize` thread occassionally reads the accumulator concurrently in the SDF implementation and it's either similar or entirely single-threaded for the unbounded source implementation. I've also split these tests out into concurrents reads and writes and isolated reads and writes, because those concurrent reads only happen every so often and occasionally overlap with writes (observed in #32921 rarely triggering `ConcurrentModificationException`). With multiple concurrent readers I had observed 10-20% improvement with padding, but the improvements are barely noticeable now if at all (on my workstation).
```
Benchmark                                                                               Mode  Cnt   Score   Error  Units
KafkaIOUtilsBenchmark.ReadAndWriteAtomic                                                avgt   15  12.948 ± 1.999  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteAtomic:atomicReadWhileWriting                         avgt   15   4.798 ± 0.200  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteAtomic:atomicWriteWhileReading                        avgt   15  21.098 ± 3.811  ns/op
KafkaIOUtilsBenchmark.ReadAndWritePaddedAtomic                                          avgt   15  13.676 ± 1.509  ns/op
KafkaIOUtilsBenchmark.ReadAndWritePaddedAtomic:paddedAtomicReadWhileWriting             avgt   15   4.598 ± 0.064  ns/op
KafkaIOUtilsBenchmark.ReadAndWritePaddedAtomic:paddedAtomicWriteWhileReading            avgt   15  22.755 ± 3.075  ns/op
KafkaIOUtilsBenchmark.ReadAndWritePlain                                                 avgt   15  15.909 ± 1.780  ns/op
KafkaIOUtilsBenchmark.ReadAndWritePlain:plainReadWhileWriting                           avgt   15   4.058 ± 0.160  ns/op
KafkaIOUtilsBenchmark.ReadAndWritePlain:plainWriteWhileReading                          avgt   15  27.760 ± 3.589  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteSynchronizedPlain                                     avgt   15  95.190 ± 1.845  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteSynchronizedPlain:synchronizedPlainReadWhileWriting   avgt   15  98.039 ± 3.973  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteSynchronizedPlain:synchronizedPlainWriteWhileReading  avgt   15  92.341 ± 2.358  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteVolatile                                              avgt   15  26.432 ± 4.415  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteVolatile:volatileReadWhileWriting                     avgt   15   3.879 ± 0.056  ns/op
KafkaIOUtilsBenchmark.ReadAndWriteVolatile:volatileWriteWhileReading                    avgt   15  48.984 ± 8.849  ns/op
KafkaIOUtilsBenchmark.ReadAtomic                                                        avgt   15   2.185 ± 0.007  ns/op
KafkaIOUtilsBenchmark.ReadPaddedAtomic                                                  avgt   15   2.193 ± 0.010  ns/op
KafkaIOUtilsBenchmark.ReadPlain                                                         avgt   15   2.206 ± 0.011  ns/op
KafkaIOUtilsBenchmark.ReadSynchronizedPlain                                             avgt   15   7.975 ± 0.571  ns/op
KafkaIOUtilsBenchmark.ReadVolatile                                                      avgt   15   2.199 ± 0.011  ns/op
KafkaIOUtilsBenchmark.WriteAtomic                                                       avgt   15   8.183 ± 0.003  ns/op
KafkaIOUtilsBenchmark.WritePaddedAtomic                                                 avgt   15   8.183 ± 0.003  ns/op
KafkaIOUtilsBenchmark.WritePlain                                                        avgt   15   9.592 ± 0.746  ns/op
KafkaIOUtilsBenchmark.WriteSynchronizedPlain                                            avgt   15  11.145 ± 1.714  ns/op
KafkaIOUtilsBenchmark.WriteVolatile                                                     avgt   15  11.967 ± 0.003  ns/op
```
I agree that it's not worth the hassle to maintain unless there's a significant upside so I've removed the layout padding from `MovingAvg` based on the results above.

sjvanrossum commented on sdks/java/io/kafka/jmh/build.gradle:
```diff
@@ -0,0 +1,31 @@
+ * limitations under the License.
+ */
+
+plugins { id 'org.apache.beam.module' }
```
Ack, I'll rewrite it.

sjvanrossum replied:
Done.

kennknowles commented on sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedReader.java:
```diff
@@ -187,13 +187,6 @@ public boolean advance() throws IOException {
           continue;
         }
 
-        long offsetGap = offset - expected; // could be > 0 when Kafka log compaction is enabled.
```
I admit I don't know this well enough to know why we tracked this or why it can be removed, even with your description in the PR. I trust your experiments, and I don't see this being a data integrity issue, though. I'd love to be educated at some point.
